File: models/mogi/lib/libmogi/Source.cc
// -*- C++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 2013-2019 parasim inc
// all rights reserved
//

// configuration
#include <portinfo>
// external
#include <cmath>
#include <pyre/journal.h>
#include <gsl/gsl_matrix.h>

// my declarations
#include "Source.h"

// make pi
const auto pi = 4*std::atan(1.0);

// meta-methods
// destructor
altar::models::mogi::Source::
~Source() {
    // make a channel
    pyre::journal::debug_t channel("mogi.source");

    // sign on
    channel
        << pyre::journal::at(__HERE__)
        << "deleting source " << this
        << pyre::journal::endl;

    // if we were handed a matrix of the locations of the observation points
    if (_locations) {
        // release it
        gsl_matrix_free(_locations);
        // tell me
        channel
            << pyre::journal::at(__HERE__)
            << "  released {_locations} matrix at" << _locations
            << pyre::journal::endl;
    }

    // if we were handed a matrix of line of sight vectors
    if (_los) {
        // release it
        gsl_matrix_free(_los);
        // tell me
        channel
            << pyre::journal::at(__HERE__)
            << "  released {_los} matrix at" << _los
            << pyre::journal::endl;
    }

    // all done
    channel
        << pyre::journal::at(__HERE__)
        << "done deleting source " << this
        << pyre::journal::endl;
}


// interface
void
altar::models::mogi::Source::
displacements(gsl_matrix_view * samples, gsl_matrix * predicted) const {
    // pull the number of samples and parameters from the shape of the {sample} matrix
    auto nSamples = samples->matrix.size1;
    auto nParameters = samples->matrix.size2;

    // go through all the samples
    for (auto sample=0; sample<nSamples; ++sample) {
        // unpack the parameters
        auto xSrc = gsl_matrix_get(&samples->matrix, sample, _xIdx);
        auto ySrc = gsl_matrix_get(&samples->matrix, sample, _yIdx);
        auto dSrc = gsl_matrix_get(&samples->matrix, sample, _dIdx);
        auto sSrc = std::pow(10, gsl_matrix_get(&samples->matrix, sample, _sIdx));

        // go through the locations
        for (auto loc=0; loc<_locations->size1; ++loc) {
            // unpack the location of the observation point
            auto xObs = gsl_matrix_get(_locations, loc, 0);
            auto yObs = gsl_matrix_get(_locations, loc, 1);

            // compute the displacement from the source to the observation point
            auto x = xSrc - xObs;
            auto y = ySrc - yObs;
            auto d = dSrc;
            // turn it into a distance
            auto R = std::sqrt(x*x + y*y + d*d);

            // compute the elastic response
            auto C = (_nu - 1) * sSrc / pi;

            // put it together
            auto CR3 = C / (R*R*R);

            // compute the components of the unit LOS vector
            auto nx = gsl_matrix_get(_los, loc, 0);
            auto ny = gsl_matrix_get(_los, loc, 1);
            auto nz = gsl_matrix_get(_los, loc, 2);

            // compute the expected displacement
            auto u = (x*nx + y*ny - d*nz) * CR3;
            // find the shift that corresponds to this observation
            auto shift = gsl_matrix_get(&samples->matrix, sample, _offsetIdx+_oids[loc]);
            // and apply it to the projected displacement
            u -= shift;

            // save
            gsl_matrix_set(predicted, sample, loc, u);
        }
    }

    // all done
    return;
}


void
altar::models::mogi::Source::
residuals(gsl_matrix * predicted) const {
    // make a channel
    pyre::journal::debug_t channel("mogi.source");

    // sign on
    channel
        << pyre::journal::at(__HERE__)
        << "computing residuals in place"
        << pyre::journal::endl;

    // unpack the number of samples and number of observations
    auto nSamples = predicted->size1;
    auto nObservations = predicted->size2;

    // go through all observations
    for (auto obs=0; obs < nObservations; ++obs) {
        // get the corresponding measurement
        auto actual = gsl_vector_get(_data, obs);
        // go though the samples
        for (auto sample=0; sample < nSamples; ++sample) {
            // get the predicted displacement
            auto pred = gsl_matrix_get(predicted, sample, obs);
            // compute the difference
            auto residual = pred - actual;
            // and store back into the matrix we were handed
            gsl_matrix_set(predicted, sample, obs, residual);
        }
    }

    // all done
    return;
}


// end of file

File: models/mogi/mogi/Source.py
# -*- python -*-
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
#
# (c) 2013-2019 parasim inc
# (c) 2010-2019 california institute of technology
# all rights reserved
#


# framework
import altar
# externals
from math import sqrt, pi as π


# declaration
class Source:
    """
    An implementation of Mogi[1958]

    The surface displacement calculation for a point pressure source in an elastic half space.
    """


    # public data
    # location
    x = 0
    y = 0
    d = 0
    # strength
    dV = 0
    # material properties
    nu = .25 # Poisson ration


    # interface
    def displacements(self, locations, los):
        """
        Compute the expected displacements from a point pressure source at a set of observation
        locations
        """
        # the location of the source
        x_src = self.x
        y_src = self.y
        d_src = self.d
        # its strength
        dV = self.dV
        # get the material properties
        nu = self.nu

        # allocate space for the result
        u = altar.vector(shape=len(locations))
        # go through each observation location
        for index, (x_obs,y_obs) in enumerate(locations):
            # compute displacements
            x = x_src - x_obs
            y = y_src - y_obs
            d = d_src
            # compute the distance to the point source
            x2 = x**2
            y2 = y**2
            d2 = d**2
            # intermediate values
            C = (nu-1) * dV/π
            R = sqrt(x2 + y2 + d2)
            CR3 = C * R**-3
            # store the expected displacement
            u[index] = x*CR3 * los[index, 0] + y*CR3 * los[index, 1] - d*CR3 * los[index,2]

        # all done
        return u


    # meta-methods
    def __init__(self, x=x, y=y, d=d, dV=dV, nu=nu, **kwds):
        # chain up
        super().__init__(**kwds)
        # store the location
        self.x = x
        self.y = y
        self.d = d
        # the strength
        self.dV = dV
        # and the material properties
        self.nu = nu
        # all done
        return


# end of file
